rules/epmark_test.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rules_test
+
+import (
+	"strings"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	. "github.com/projectcalico/felix/rules"
+)
+
+func init() {
+
+	ErrMark := uint32(0)
+
+	DescribeTable("EndpointMarkMapper initialization",
+		func(mask uint32) {
+			epmm := NewEndpointMarkMapperWithShim(mask, &mockHash32{})
+			Expect(epmm.GetMask()).To(Equal(mask))
+		},
+		Entry("should initialise with one bit", uint32(0x10)),
+		Entry("should initialise with some bits", uint32(0x123f)),
+		Entry("should initialise with max bits", uint32(0xffffffff)),
+	)
+
+	Describe("EndpointMarkMapper allocation/release", func() {
+
+		var epmm EndpointMarkMapper
+		BeforeEach(func() {
+			epmm = NewEndpointMarkMapperWithShim(0x700, &mockHash32{})
+		})
+		DescribeTable("EndpointMarkMapper allocation",
+			func(eps []string, expected []uint32) {
+				result := []uint32{}
+				for _, ep := range eps {
+					if strings.HasPrefix(ep, "x") {
+						epmm.ReleaseEndpointMark(strings.TrimPrefix(ep,"x"))
+					} else {
+						// if error, function return 0 which match ErrMark.
+						mark, _ := epmm.GetEndpointMark(ep)
+						result = append(result, mark)
+					}
+				}
+				Expect(result).To(Equal(expected))
+			},
+			Entry("should allocate sequentially and repeat",
+				[]string{"cali1", "cali2", "cali3", "cali1", "cali2", "cali7"},
+				[]uint32{0x100, 0x200, 0x300, 0x100, 0x200, 0x700}),
+			Entry("should allocate with collision and fail correctly",
+				[]string{"cali1", "cali6", "cali3", "cali11", "cali22", "cali33", "cali8", "cali9"},
+				[]uint32{0x100, 0x600, 0x300, 0x200, 0x400, 0x500, 0x700, ErrMark}),
+			Entry("should allocate/release with collision",
+				[]string{"cali1", "cali6", "cali3", "xcali1", "xcali2", "cali33", "cali11", "cali66"},
+				[]uint32{0x100, 0x600, 0x300, 0x400, 0x100, 0x700}),
+		)
+	})
+
```

Comment:
Should add tests that use all the marks, then some that fail because there aren't any free marks.  Then free some up and check that allocating more works.

Reply:
DONE.